src/refiners/foundationals/latent_diffusion/freeu.py

```diff
@@ -0,0 +1,86 @@
+import torch
+from torch.fft import fftn, fftshift, ifftn, ifftshift
+from typing import cast, Iterable
+from refiners.fluxion.layers.module import Module
+import refiners.fluxion.layers as fl
+from refiners.foundationals.latent_diffusion.stable_diffusion_1.unet import ResidualConcatenator
+
+# Ported from https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/torch_utils.py
+def fourier_filter(x_in, threshold, scale):
+    """Fourier filter as introduced in FreeU (https://arxiv.org/abs/2309.11497).
+
+    This version of the method comes from here:
+    https://github.com/huggingface/diffusers/pull/5164#issuecomment-1732638706
+    """
+    x = x_in
+    B, C, H, W = x.shape
+
+    # Non-power of 2 images must be float32
+    if (W & (W - 1)) != 0 or (H & (H - 1)) != 0:
+        x = x.to(dtype=torch.float32)
+
+    # FFT
+    x_freq = fftn(x, dim=(-2, -1))
+    x_freq = fftshift(x_freq, dim=(-2, -1))
+
+    B, C, H, W = x_freq.shape
+    mask = torch.ones((B, C, H, W), device=x.device)
+
+    crow, ccol = H // 2, W // 2
+    mask[..., crow - threshold : crow + threshold, ccol - threshold : ccol + threshold] = scale
+    x_freq = x_freq * mask
+
+    # IFFT
+    x_freq = ifftshift(x_freq, dim=(-2, -1))
+    x_filtered = ifftn(x_freq, dim=(-2, -1)).real
+
+    return x_filtered.to(dtype=x_in.dtype)
+
+# Ported from https://github.com/huggingface/diffusers/blob/main/src/diffusers/utils/torch_utils.py
+class FreeUHiddenStates(Module):
+    def __init__(self, b: float):
+        super().__init__()
+        self.b = b
+    def forward(self, hidden_states):
+        num_half_channels = hidden_states.shape[1] // 2
+        hidden_states[:, :num_half_channels] = hidden_states[:, :num_half_channels] * self.b
+        return hidden_states
+class FreeUSkipConnections(Module):
+    def __init__(self, s: float):
+        super().__init__()
+        self.s = s
+    def forward(self, hidden_states):
+        hidden_states = fourier_filter(hidden_states, threshold=1, scale=self.s)
+        return hidden_states
+
+class FreeUAdapter(fl.Chain):
+    def __init__(self, n: int, b: float, s: float) -> None:
+        self.n = n
+
+        super().__init__(
+            fl.Concatenate(
+                FreeUHiddenStates(b),
+                fl.Chain(
+                    fl.UseContext(context="unet", key="residuals").compose(lambda residuals: residuals[self.n]),
+                    FreeUSkipConnections(s)
+                ),
+                dim=1,
+            ),
+        )
+
+class SD1UnetFreeUAdapter:
+    def __init__(self, unet, bs: [float], ss: [float]):
+        assert len(bs) == len(ss)
+        self.unet = unet
```

Comment:
For composability reasons, we always realise adapters as a fl.Chain with the Adapter class.

You can take example on the SelfAttentionGuidance which is also a weightless adapter on top of SDUNet: https://github.com/finegrain-ai/refiners/blob/main/src/refiners/foundationals/latent_diffusion/self_attention_guidance.py#L67

Reply:
Thanks! Fixing now.

Reply:
I think fixed. Let me know what you think